### PR TITLE
Update regex used to remove ANSI escape sequences to be more specific to decoration and hyperlinks

### LINKED ISSUE #16730: out-file producing weird characters
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When using out-file with or without specifying any encoding the file has some weird characters at the beginning of the file

[32;1m NPM(K)    PM(M)      WS(M)     CPU(s)      Id  SI ProcessName[0m
[32;1m ------    -----      -----     ------      --  -- -----------[0m

I've tried this on different computers all using ps 7.2.1, windows 10, server 2016 etc. When using the same get-process |out-file with powershell 5.1 it works as expected without those characters.  

### Expected behavior

```console
NPM(K)    PM(M)      WS(M)     CPU(s)      Id  SI ProcessName
```

### Actual behavior

```console
[32;1m NPM(K)    PM(M)      WS(M)     CPU(s)      Id  SI ProcessName[0m
```

### Error details

_No response_

### Environment data

```powershell
7.2.1
```

### Visuals

image

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

The previous regex to remove ANSI escape sequences was too broad and would remove content that appeared to be ANSI but was not.  The change is to have specific regex targeting colors/decoration and CSI ANSI escape sequences (generated by PSReadLine) which are ones used by PS7 and PSReadLine currently.

Hyperlink isn't supported here as no default formatting emits it and the way the escape sequence works it can be too broad.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/16730

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/FormatAndOutput/common/PSStyle.cs
Comment (iSazonov) [inline src/System.Management.Automation/FormatAndOutput/common/PSStyle.cs:575]: Why and where it was changed to Ansi? :-)

Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/PSStyle.cs:575]: This was a temporary change to address https://github.com/PowerShell/Announcements/issues/28.  This PR is trying to fix the underlying problem so we can change it back.

**Comment by iSazonov** [comment]
Re-assign PR to security WG.

**Comment by iSazonov** [comment]
@SteveL-MSFT Sorry if I already said this but I feel we fall in unlimited problems and fixes. I would personally roll back this kind of coloring (it's a far cry from the spirit of PowerShell) and invest in Formatting system enhancement or better in  something like PSMore.

**Comment by SteveL-MSFT** [comment]
@iSazonov PSMore is solving a different problem than PSStyle.  The two are not mutually exclusive.  ANSI/VT escape sequences are the industry standard way of emitting text decoration and works across ssh, xterm.js, Windows, Linux, macOS, native commands, etc... For PSMore to emit color, it will have to resort to ANSI/VT escape sequences anyways and we would still have this situation.  Even if table/list formats don't use ANSI/VT, arbitrary scripts and native commands can and do emit ANSI/VT.

**Comment by iSazonov** [comment]
> ANSI/VT escape sequences are the industry standard

I don't say about rejecting ESs - they are _only_ implementation details. _If we will continue on the way we **obviously** fall in re-implementing whole Windows Terminal in PowerShell!_ - there is no other way to ensure full processing of ESs in all scenarios.

> arbitrary **scripts** and native commands can and do emit ANSI/VT.

This is my main frustration because we are provoking users to use ESs when the spirit and idea of PowerShell has always been to do **magical** things.

> PSMore is solving a different problem than PSStyle.

I mentioned PSMore as a way to develop without breaking the existing Formatting System.
This would allow us to find right design and implement it step by step.
I'm sure we need (1) to delegate the details to the console and the terminal and (2) use higher abstractions in the formatting system and even more so in the PowerShell language itself.

Sorry for noisy :-)

**Comment by SteveL-MSFT** [comment]
@iSazonov I fully agree that the current formatting system needs to be revamped, but again, that's independent of ESC seqs.  We can always adopt something like spectre.console and have something like "{red}hello {blue}there", although the benefit of "$($psstyle.foreground.red)hello" is you get tab completion.  Anyways, I think this discussion is separate from this PR.

**Comment by iSazonov** [comment]
@SteveL-MSFT All you say is about **string** formatting. My proposal is to follow PowerShell spirit and invest in **object** formatting first.
String formatting fundamentally leads us to re-implement Windows Terminal in PowerShell.
Object-based formatting works well for years and could be improved for modern terminals (not easy work!). (As part of this problem I proposed Native cmdlet class #13428 which could universally collect all meta data about native command and allow PowerShell engine reliable communicate with the native app (as with regular cmdlet) including smart re-incoding input and output.)

**Review by PaulHigin — APPROVED** [review]
This looks Ok to me.  But still need @TravisEZ13 look at the regex.

**Review by daxian-dbw — APPROVED** [review]

**Comment by TravisEz13** [comment]
@PowerShell/powershell-maintainers  Let's get at least one preview release before we approve backport

**Comment by daxian-dbw** [comment]
@SteveL-MSFT Just noticed this PR changed the default `OutputRendering` back to `Host`. Is that intentional? I thought we changed it to `ANSI` for a security issue. Is that security issue resolved after using the new regex for detecting ANSI sequences, so we can change back to `Host`? /cc @TravisEz13

**Comment by ImportTaste** [comment]
> @SteveL-MSFT Just noticed this PR changed the default `OutputRendering` back to `Host`. Is that intentional? I thought we changed it to `ANSI` for a security issue. Is that security issue resolved after using the new regex for detecting ANSI sequences, so we can change back to `Host`? /cc @TravisEz13

daxian, were you able to track down someone to answer this question for you, in another PR/Issue/etc perhaps? I'm curious to the answer myself.

**Comment by daxian-dbw** [comment]
My chat with @TravisEz13 suggested this is an intentional change, and `Host` will be used as the default unless another vulnerability is found with the regex.

**Comment by adityapatwardhan** [comment]
@PowerShell/powershell-maintainers should discuss whether this should be backported.

**Pull request merged**
